mediapipe/graphs/instantmotiontracking/calculators/matrices_manager.cc

```diff
@@ -0,0 +1,221 @@
+  }
+  if (cc->Outputs().HasTag(kModelMatricesTag)) {
+    cc->Outputs().Tag(kModelMatricesTag).Set<TimedModelMatrixProtoList>();
+  }
+
+  return ::mediapipe::OkStatus();
+}
+
+::mediapipe::Status MatricesManagerCalculator::Open(CalculatorContext* cc) {
+  cc->SetOffset(TimestampDiff(0));
+  return ::mediapipe::OkStatus();
+}
+
+::mediapipe::Status MatricesManagerCalculator::Process(CalculatorContext* cc) {
+  auto model_matrices = std::make_unique<TimedModelMatrixProtoList>();
+  TimedModelMatrixProtoList* model_matrix_list = model_matrices.get();
+  model_matrix_list->clear_model_matrix();
+
+  const std::vector<UserRotation> user_rotation_data =
+      cc->Inputs().Tag(kUserRotationsTag).Get<std::vector<UserRotation>>();
+
+  const std::vector<UserScaling> user_scaling_data =
+        cc->Inputs().Tag(kUserScalingsTag).Get<std::vector<UserScaling>>();
+
+  const std::vector<Anchor> translation_data =
+      cc->Inputs()
+          .Tag(kFinalTranslationsTag)
+          .Get<std::vector<Anchor>>();
+
+  // Device IMU Data definitions
+  const float roll = cc->Inputs().Tag(kIMUDataTag).Get<float[]>()[0];
+  const float pitch = cc->Inputs().Tag(kIMUDataTag).Get<float[]>()[1];
+  const float yaw = cc->Inputs().Tag(kIMUDataTag).Get<float[]>()[2];
+
+  for (Anchor anchor : translation_data) {
+    int id = anchor.sticker_id;
+
+    TimedModelMatrixProto* model_matrix =
+        model_matrix_list->add_model_matrix();
+    model_matrix->set_id(id);
+
+    float rotation = getUserRotation(user_rotation_data, id);
+    float scaling = getUserScaling(user_scaling_data, id);
+
+    Eigen::Vector3f translation_vector = generateTranslationVector(anchor, scaling);
+    Eigen::Matrix3f rotation_submatrix = generateRotationSubmatrix(roll, pitch, yaw, rotation);
+
+    Matrix4fRM mvp_matrix = generateEigenModelMatrix(translation_vector, rotation_submatrix);
+
+    // The generated model matrix must be mapped to TimedModelMatrixProto (col-wise)
+    for (int i = 0; i < mvp_matrix.rows(); ++i) {
+      for (int j = 0; j < mvp_matrix.cols(); ++j) {
+        model_matrix->add_matrix_entries(mvp_matrix(j, i));
+      }
+    }
+  }
+
+  cc->Outputs()
+      .Tag(kModelMatricesTag)
+      .Add(model_matrices.release(), cc->InputTimestamp());
+
+  return ::mediapipe::OkStatus();
+}
+
+// Using the tracked translation data, generate a vector for MVP translation
+Eigen::Vector3f MatricesManagerCalculator::generateTranslationVector(Anchor tracked_anchor, float user_scaling_increment) {
+  // Convert from normalized [0.0-1.0] to openGL on-screen coordinates
+  float z = (initial_z_ + user_scaling_increment) * tracked_anchor.z;
+
+  float y_minimum = z * (tan(vertical_fov_radians_ /2));
+  // Minimum y value appearing on screen at z distance
+  float x_minimum = y_minimum * (1.0/aspect_ratio_);
+  // Minimum x value appearing on screen at z distance
+
+  float x = (tracked_anchor.x * (-x_minimum - x_minimum)) + x_minimum;
```

Comment:
This is an interesting approach-- I'd need to think more about how this matches up to the equations you get by back-solving the projection matrix application, but I'm wondering if this might add a little extra warping of the tracking speed across the screen... we should definitely add a TODO to investigate possible differences there.